File: backend/app/models/autotag.py
"""Pydantic models for audio auto-tagging (Apple Music / iTunes).

The contract mirrored by the frontend's `src/types/autotag.ts`. The flow is
identify/search → (user picks a version + edits) → apply; nothing is written to
the file until `apply`.
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class TagCandidate(BaseModel):
    """One catalogue match (an Apple Music track release) to tag a file with."""

    title: str = Field(..., description="Track title.")
    artist: str = Field(..., description="Track artist(s).")
    album: str | None = Field(default=None, description="Album / collection name.")
    year: str | None = Field(default=None, description="Release year (YYYY).")
    track_number: int | None = Field(default=None, description="Track number.")
    cover_url: str | None = Field(default=None, description="Cover art URL.")


class IdentifyRequest(BaseModel):
    """Look up matches for an already-downloaded audio file (by its filename)."""

    path: str = Field(..., description="Absolute path to the audio file.")


class SearchRequest(BaseModel):
    """Manual catalogue search."""

    artist: str = Field(default="", description="Artist to search for.")
    title: str = Field(..., description="Track title to search for.")


class CandidateList(BaseModel):
    """Matches returned by identify/search (best first; empty = nothing found)."""

    results: list[TagCandidate] = Field(default_factory=list)


class ApplyRequest(BaseModel):
    """Write the (possibly user-edited) tags + cover into the file."""

    path: str = Field(..., description="Absolute path to the audio file.")
    title: str | None = Field(default=None)
    artist: str | None = Field(default=None)
    album: str | None = Field(default=None)
    year: str | None = Field(default=None)
    track_number: int | None = Field(default=None)
    cover_url: str | None = Field(default=None, description="Cover art to embed.")
    embed_lyrics: bool | None = Field(
        default=None,
        description="Embed lyrics for this track (None = use the global setting).",
    )


class ApplyResponse(BaseModel):
    ok: bool = Field(..., description="Whether the tags were written.")
    embedded_cover: bool = Field(
        default=False, description="Whether cover art was embedded."
    )


class LyricsRequest(BaseModel):
    """Preview-time lyrics lookup for the auto-tag card."""

    title: str = Field(..., description="Track title.")
    artist: str = Field(default="", description="Track artist(s).")
    album: str | None = Field(default=None)
    path: str | None = Field(
        default=None, description="Audio file, to probe its duration for a match."
    )


class LyricsResult(BaseModel):
    found: bool = Field(..., description="Whether any lyrics were found.")
    instrumental: bool = Field(default=False)
    has_synced: bool = Field(default=False, description="A timed .lrc is available.")
    plain: str | None = Field(default=None, description="The full plain lyrics.")

File: src/types/autotag.ts
/**
 * Audio auto-tagging types — mirror of `backend/app/models/autotag.py`.
 *
 * Metadata + cover art come from the Apple Music (iTunes) catalogue. Flow:
 * identify/search → (user picks a version + edits) → apply. Nothing is written
 * to the file until `apply`.
 */

/** One catalogue match (an Apple Music track release). */
export interface TagCandidate {
  title: string;
  artist: string;
  album: string | null;
  year: string | null;
  genre: string | null;
  track_number: number | null;
  cover_url: string | null;
}

/** Matches from identify/search (best first; empty = nothing found). */
export interface CandidateList {
  results: TagCandidate[];
}

/** The (possibly user-edited) tags to write into the file. */
export interface ApplyRequest {
  path: string;
  title?: string | null;
  artist?: string | null;
  album?: string | null;
  year?: string | null;
  genre?: string | null;
  track_number?: number | null;
  cover_url?: string | null;
  /** Embed lyrics for this track (omitted = use the global setting). */
  embed_lyrics?: boolean | null;
}

export interface ApplyResponse {
  ok: boolean;
  embedded_cover: boolean;
}

/** Preview-time lyrics availability for the auto-tag card. */
export interface LyricsResult {
  found: boolean;
  instrumental: boolean;
  has_synced: boolean;
  /** The full plain lyrics (shown in the "view lyrics" popup). */
  plain: string | null;
}
